### PR TITLE
Calling HashMap constructors with empty entries returns "untouched" empty HashMap

### PR DESCRIPTION
**Pull request by AlexGeb**
## Type

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Make `HashMap.make` and `HashMap.fromIterable` return an "untouched" empty `HashMap` when called with empty entries.

Before this change, the below assertion failed because even when the given `Iterable` is empty, we would create an `HashMap` with one edit.

```typescript
deepStrictEqual(HM.fromIterable([]), HM.empty())
```
Screenshot 2024-09-25 at 20 54 31

## Related

- Related Issue #
- Closes #

### CODE REVIEW

File: packages/effect/src/internal/hashMap.ts [hunk 1]
```diff
@@ -191,11 +191,12 @@ export const make = <Entries extends ReadonlyArray<readonly [any, any]>>(
...
 export const fromIterable = <K, V>(entries: Iterable<readonly [K, V]>): HM.HashMap<K, V> => {
-  const map = beginMutation(empty<K, V>())
+  const emptyMap = empty<K, V>()
+  const map = beginMutation(emptyMap)
   for (const entry of entries) {
     set(map, entry[0], entry[1])
   }
-  return endMutation(map)
+  return isEmpty(map) ? emptyMap : endMutation(map)
 }
```
Comment (AlexGeb) [inline packages/effect/src/internal/hashMap.ts:200]: We could also do:
```ts
/** @internal */
export const fromIterable = <K, V>(entries: Iterable<readonly [K, V]>): HM.HashMap<K, V> => {
  const emptyMap = empty<K, V>()
  const map = beginMutation(emptyMap)
  for (const entry of entries) {
    set(map, entry[0], entry[1])
  }
  endMutation(map)
  ;(map as HashMapImpl<K, V>)._edit = 0
  return map
}
```

**Comment by mikearnaldi** [comment]
It's not a bug, deepStrictEqual doesn't generally work to compare HashMap values. No point special casing it for empty maps

**Comment by AlexGeb** [comment]
> It's not a bug, deepStrictEqual doesn't generally work to compare HashMap values. No point special casing it for empty maps

Ok, we might not need deep strict equality for comparing HashMap (it's the default in our codebase, every tests we write are using deep strict equality, we never challenged this before, it's maybe the time to do it).

BUT there is still the fact that the `_edit` property is equal to 1 whereas we just called a constructor. I would expect that calling an HashMap constructor would give me an "untouched" HashMap..

**Comment by mikearnaldi** [comment]
Those are to be considered internals, a consumer of the module should not rely on internals. BTW what function are you using? In the vitest custom matchers we add rules to compare effect modules so you should have a positive result in comparing two empty maps regardless of the internals

**Comment by AlexGeb** [comment]
> Those are to be considered internals, a consumer of the module should not rely on internals. BTW what function are you using? In the vitest custom matchers we add rules to compare effect modules so you should have a positive result in comparing two empty maps regardless of the internals

Ah we use the `expect().toStrictEqual` but with expect being automatically injected by vitest in test files (we do not import anything related to tests, its globally scoped for test files).
We only import `it` from '@effect/vitest' when we need test effects.

I think that what would be great for us would be to replace vitest default methods by the one from `@effect/vitest` (this way we can keep same devX)

**Comment by mikearnaldi** [comment]
> > Those are to be considered internals, a consumer of the module should not rely on internals. BTW what function are you using? In the vitest custom matchers we add rules to compare effect modules so you should have a positive result in comparing two empty maps regardless of the internals
> 
> Ah we use the `expect().toStrictEqual` but with expect being automatically injected by vitest in test files (we do not import anything related to tests, its globally scoped for test files). We only import `it` from '@effect/vitest' when we need test effects.
> 
> I think that what would be great for us would be to replace vitest default methods by the one from `@effect/vitest` (this way we can keep same devX)

it's fine to have vitest auto-inject the methods `expect` plugins are a default feature, you need to register them though, like we do in: https://github.com/Effect-TS/effect/blob/main/setupTests.ts

**Comment by AlexGeb** [comment]
> it's fine to have vitest auto-inject the methods expect plugins are a default feature, you need to register them though, like we do in: https://github.com/Effect-TS/effect/blob/main/setupTests.ts

Awesome, I didn't knew about it! I'm closing this then, thank you for your answers

**Pull request closed without merging**